lib/stat_tracker.rb

```diff
@@ -14,34 +21,309 @@ def self.from_csv(data_files)
+      game.home_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_away_goals_scored(team_id)
+    away_games_filtered_by_team(team_id).min_by do |game|
+      game.away_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_goals_scored(team_id)
+    if fewest_home_goals_scored(team_id) < fewest_away_goals_scored(team_id)
+      fewest_home_goals_scored(team_id)
+    else
+      fewest_away_goals_scored(team_id)
+    end
+  end
+
+  def all_games_played_by_team(team_id)
+    game_collection.all.find_all do |game|
+      game.home_team_id == team_id || game.away_team_id == team_id
+    end
+  end
+
+  def opponents(team_id)
+    # We want:
+                # {
+                #   "some_opponent_id" =>
+                #     {
+                #       "won" => times_some_opponent_won_against_my_team,
+                #       "lost" => times_some_opponent_lost_against_my_team,
+                #       "tied" => times_some_opoonent_tied_against_my_team
+                #     }
+                # }
+    acc = Hash.new
+    all_games_played_by_team(team_id).each do |game|
+      if game.away_team_id == team_id
+        opponent_id = game.home_team_id
+      else
+        opponent_id = game.away_team_id
+      end
+
+      # if the block has been iterated through before, then set old_value won/lost/tied to that previous value, and if it has not been iterated through then set old value won/lost/tied to 0
+
+      if acc[opponent_id].nil?
+        old_value_won = 0.0
+        old_value_lost = 0.0
+        old_value_tied = 0.0
+      else
+        old_value_won = acc[opponent_id][:won]
+        old_value_lost = acc[opponent_id][:lost]
+        old_value_tied = acc[opponent_id][:tied]
+      end
+
+      if opponent_id == game.winning_team_id
+        old_value_won += 1.0
+      elsif opponent_id == game.losing_team_id
+        old_value_lost += 1.0
+      else
+        old_value_tied += 1.0
       end
+
+      acc[opponent_id] = {:won => old_value_won, :lost => old_value_lost, :tied => old_value_tied}
     end
-    lowest_score
+    acc
   end
+
+  def opponent_win_percentages(team_id)
+    # We want: {"opponent_id" => win_percentage, "another_opponent_id" => their_win_percentage}
+    acc = {}
+    result = opponents(team_id).each do |opponent, data|
+      acc[opponent] = data[:won]/data.values.sum.to_f
+    end
+    acc
+  end
+
+  def most_won_against_opponent(team_id)
+    opponent_win_percentages(team_id).min_by do |opponent_id, win_rate|
+      win_rate
+    end.first
+  end
+
+  def favorite_opponent(team_id)
+    team_collection.all.find do |team|
+      most_won_against_opponent(team_id) == team.team_id
+    end.team_name
+  end
+
+  def most_lost_against_opponent(team_id)
+    opponent_win_percentages(team_id).max_by do |opponent_id, win_rate|
+      win_rate
+    end.first
+  end
+
+  def rival(team_id)
+    acc = team_collection.all.find do |team|
+      most_lost_against_opponent(team_id) == team.team_id
+    end.team_name
+  end
+
 end
```

Comment:
Such a great use of helper methods Dan!!! Really clear what is going on in the code and the implementation of such across all the methods was cohesive and clever!